drivers/flash/flash_stm32wb0x.c

```diff
@@ -0,0 +1,475 @@
+}
+
+/**
+ * Driver subsystem API implementation
+ */
+int flash_wb0x_read(const struct device *dev, off_t offset,
+			void *buffer, size_t len)
+{
+	if (!len) {
+		return 0;
+	}
+
+	if (!is_valid_flash_range(dev, offset, len)) {
+		return -EINVAL;
+	}
+
+	const uint8_t *flash_base = ((void *)DT_REG_ADDR(DT_INST(0, st_stm32_nv_flash)));
+
+	memcpy(buffer, flash_base + (uint32_t)offset, len);
+
+	return 0;
+}
+
+int flash_wb0x_write(const struct device *dev, off_t offset,
+			const void *buffer, size_t len)
+{
+	struct flash_wb0x_data *data = dev->data;
+	int res;
+
+	if (!len) {
+		return 0;
+	}
+
+	if (!is_writeable_flash_range(dev, offset, len)) {
+		return -EINVAL;
+	}
+
+	/* Acquire driver lock */
+	res = k_sem_take(&data->write_lock, K_NO_WAIT);
+	if (res < 0) {
+		return res;
+	}
+
+	const uint32_t start_word = (uint32_t)offset / WORD_SIZE;
+	const uint32_t num_words = len / WORD_SIZE;
+
+	res = write_word_range(buffer, start_word, num_words);
+
+	/* Release driver lock */
+	k_sem_give(&data->write_lock);
+
+	return res;
+}
+
+int flash_wb0x_erase(const struct device *dev, off_t offset, size_t size)
+{
+	struct flash_wb0x_data *data = dev->data;
+	int res;
+
+	if (!size) {
+		return 0;
+	}
+
+	if (!is_erasable_flash_range(dev, offset, size)) {
+		return -EINVAL;
+	}
+
+	/* Acquire driver lock */
+	res = k_sem_take(&data->write_lock, K_NO_WAIT);
+	if (res < 0) {
+		return res;
+	}
+
+	const uint32_t start_page = (uint32_t)offset / ERASE_BLOCK_SIZE;
+	const uint32_t page_count = size / ERASE_BLOCK_SIZE;
+
+	res = erase_page_range(start_page, page_count);
+
+	/* Release driver lock */
+	k_sem_give(&data->write_lock);
+
+	return res;
+}
+
+const struct flash_parameters *flash_wb0x_get_parameters(
+					const struct device *dev)
+{
+	static const struct flash_parameters fp = {
+		.write_block_size = WRITE_BLOCK_SIZE,
+		.erase_value = 0xff,
+	};
+
+	return &fp;
+}
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+void flash_wb0x_pages_layout(const struct device *dev,
+				const struct flash_pages_layout **layout,
+				size_t *layout_size)
+{
+	/**
+	 * STM32WB0 flash: single bank, 2KiB pages
+	 * (the number of pages depends on MCU)
+	 */
+	static const struct flash_pages_layout fpl[] = {{
+		.pages_count = PAGES_IN_FLASH,
+		.pages_size = FLASH_PAGE_SIZE
+	}};
+
+	*layout = fpl;
+	*layout_size = ARRAY_SIZE(fpl);
+}
+#endif /* CONFIG_FLASH_PAGE_LAYOUT */
+
+#if defined(CONFIG_FLASH_EX_OP_ENABLED)
+int flash_wb0x_ex_op(const struct device *dev, uint16_t code,
+			const uintptr_t in, void *out)
+{
+	/* Extended operations are not supported */
+	return -ENOTSUP;
+}
+#endif /* CONFIG_FLASH_EX_OP_ENABLED */
+
+static const struct flash_driver_api flash_wb0x_api = {
+	.erase = flash_wb0x_erase,
+	.write = flash_wb0x_write,
+	.read = flash_wb0x_read,
+	.get_parameters = flash_wb0x_get_parameters,
+#ifdef CONFIG_FLASH_PAGE_LAYOUT
+	.page_layout = flash_wb0x_pages_layout,
+#endif
+#ifdef CONFIG_FLASH_EX_OP_ENABLED
+	.ex_op = flash_wb0x_ex_op,
+#endif
```

Comment:
API already returns `-ENOTSUP` if not defined

Reply:
Good catch. Removing this useless stub.